programs/vest_api/src/vest_state.rs

```diff
@@ -0,0 +1,110 @@
+//! vest state
+use crate::vest_schedule::create_vesting_schedule;
+use bincode::{self, deserialize, serialize_into};
+use chrono::prelude::*;
+use chrono::{
+    prelude::{DateTime, TimeZone, Utc},
+    serde::ts_seconds,
+};
+use serde_derive::{Deserialize, Serialize};
+use solana_sdk::{account::Account, instruction::InstructionError, pubkey::Pubkey};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct VestState {
+    /// The address authorized to terminate this contract with a signed Terminate instruction
+    pub terminator_pubkey: Pubkey,
+
+    /// The address authorized to redeem vested tokens
+    pub payee_pubkey: Pubkey,
+
+    /// The day from which the vesting contract begins
+    #[serde(with = "ts_seconds")]
+    pub start_dt: DateTime<Utc>,
+
+    /// Address of an account containing a trusted date, used to drive the vesting schedule
+    pub date_pubkey: Pubkey,
+
+    /// The number of lamports to send the payee if the schedule completes
+    pub lamports: u64,
+
+    /// The number of lamports the payee has already redeemed
+    pub redeemed_lamports: u64,
+}
+
+impl Default for VestState {
+    fn default() -> Self {
+        Self {
+            terminator_pubkey: Pubkey::default(),
+            payee_pubkey: Pubkey::default(),
+            start_dt: Utc.timestamp(0, 0),
+            date_pubkey: Pubkey::default(),
+            lamports: 0,
+            redeemed_lamports: 0,
+        }
+    }
+}
+
+impl VestState {
+    pub fn serialize(&self, output: &mut [u8]) -> Result<(), InstructionError> {
+        serialize_into(output, self).map_err(|_| InstructionError::AccountDataTooSmall)
+    }
+
+    pub fn deserialize(input: &[u8]) -> Result<Self, InstructionError> {
+        deserialize(input).map_err(|_| InstructionError::InvalidAccountData)
+    }
+
+    /// Redeem vested tokens.
+    pub fn redeem_tokens(
+        &mut self,
+        current_dt: Date<Utc>,
+        contract_account: &mut Account,
+        payee_account: &mut Account,
+    ) {
+        let schedule = create_vesting_schedule(self.start_dt.date(), self.lamports);
+
+        let vested_lamports = schedule
+            .into_iter()
+            .take_while(|(dt, _)| *dt <= current_dt)
+            .map(|(_, lamports)| lamports)
+            .sum::<u64>();
+
+        let redeemable_lamports = vested_lamports.saturating_sub(self.redeemed_lamports);
+
+        contract_account.lamports -= redeemable_lamports;
+        payee_account.lamports += redeemable_lamports;
+
+        self.redeemed_lamports += redeemable_lamports;
+    }
+
+    /// Terminate the contract and return all tokens to the given pubkey.
+    pub fn terminate(&mut self, contract_account: &mut Account, payee_account: &mut Account) {
```

Comment:
What's the use of `terminate`? It seems the payee account is getting all the lamports on terminate.

Reply:
On reading it again, I think payee account in terminate might be different than payee account in redeem_token. So it's just a way to drain the contract account to some account... If so, disregard my question above.